src/detect-parse.c

```diff
@@ -353,42 +386,81 @@ void SigTableApplyStrictCommandlineOption(const char *str)
 {
-    if (list > 0 && (uint32_t)list >= s->init_data->smlists_array_size)
-    {
-        uint32_t old_size = s->init_data->smlists_array_size;
-        uint32_t new_size = (uint32_t)list + 1;
-        void *ptr = SCRealloc(s->init_data->smlists, (new_size * sizeof(SigMatch *)));
-        if (ptr == NULL)
-            abort();
-        s->init_data->smlists = ptr;
-        ptr = SCRealloc(s->init_data->smlists_tail, (new_size * sizeof(SigMatch *)));
-        if (ptr == NULL)
-            abort();
-        s->init_data->smlists_tail = ptr;
-        for (uint32_t i = old_size; i < new_size; i++) {
-            s->init_data->smlists[i] = NULL;
-            s->init_data->smlists_tail[i] = NULL;
-        }
-        s->init_data->smlists_array_size = new_size;
+    if (new->type == DETECT_CONTENT) {
+        s->init_data->max_content_list_id = MAX(s->init_data->max_content_list_id, (uint32_t)list);
     }
 
-    if (s->init_data->smlists[list] == NULL) {
-        s->init_data->smlists[list] = new;
-        s->init_data->smlists_tail[list] = new;
-        new->next = NULL;
-        new->prev = NULL;
+    SCLogDebug("s:%p new:%p list:%d: %s, s->init_data->list_set %s s->init_data->list %d", s, new,
+            list, sigmatch_table[new->type].name, BOOL2STR(s->init_data->list_set),
+            s->init_data->list);
+
+    if (list < DETECT_SM_LIST_MAX) {
+        if (s->init_data->smlists[list] == NULL) {
+            s->init_data->smlists[list] = new;
+            s->init_data->smlists_tail[list] = new;
+            new->next = NULL;
+            new->prev = NULL;
+        } else {
+            SigMatch *cur = s->init_data->smlists_tail[list];
+            cur->next = new;
+            new->prev = cur;
+            new->next = NULL;
+            s->init_data->smlists_tail[list] = new;
+        }
+        new->idx = s->init_data->sm_cnt;
+        s->init_data->sm_cnt++;
+
     } else {
-        SigMatch *cur = s->init_data->smlists_tail[list];
-        cur->next = new;
-        new->prev = cur;
-        new->next = NULL;
-        s->init_data->smlists_tail[list] = new;
-    }
+        /* app-layer-events (and possibly others?) can get here w/o a "list"
+         * already set up. */
 
-    new->idx = s->init_data->sm_cnt;
-    s->init_data->sm_cnt++;
+        /* unset any existing list if it isn't the same as the new */
+        if (s->init_data->list != DETECT_SM_LIST_NOTSET && list != s->init_data->list) {
+            SCLogDebug("reset: list %d != s->init_data->list %d", list, s->init_data->list);
+            s->init_data->list = DETECT_SM_LIST_NOTSET;
+        }
+        if ((s->init_data->curbuf != NULL && (int)s->init_data->curbuf->id != list) ||
+                s->init_data->curbuf == NULL) {
+            if (s->init_data->buffer_index + 1 == s->init_data->buffers_size) {
+                void *ptr = SCRealloc(s->init_data->buffers, s->init_data->buffers_size + 8);
+                BUG_ON(ptr == 0);
```

Comment:
error handle
also: this code looks similar to other reallocs of this array, so reuse?